Re-implement whole Activation Key test module. Remove all stubs

oshtaier opened the pull request:
Closes #3176

```
nosetests tests/foreman/cli/test_activationkey.py
.F...F....SS.E..........................S.......
======================================================================
ERROR: @Test: Positive content override
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/oshtaier/Desktop/robottelo/robottelo/decorators.py", line 445, in wrapper_func
    return func(*args, **kwargs)
  File "/home/oshtaier/Desktop/robottelo/tests/foreman/cli/test_activationkey.py", line 1145, in test_positive_content_override
    self.assertEqual(content[0]['enabled?'], override_value)
KeyError: 'enabled?'
======================================================================
FAIL: @Test: Create Activation key with invalid Description
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/oshtaier/Desktop/robottelo/robottelo/decorators.py", line 224, in wrapper
    return func(*args, **kwargs)
  File "/home/oshtaier/Desktop/robottelo/robottelo/decorators.py", line 445, in wrapper_func
    return func(*args, **kwargs)
  File "/home/oshtaier/Desktop/robottelo/tests/foreman/cli/test_activationkey.py", line 219, in test_negative_create_with_description
    {u'description': gen_string('alpha', 1001)})
AssertionError: CLIFactoryError not raised
======================================================================
FAIL: @Test: Try to update Activation Key using invalid value for its
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/oshtaier/Desktop/robottelo/robottelo/decorators.py", line 445, in wrapper_func
    return func(*args, **kwargs)
  File "/home/oshtaier/Desktop/robottelo/tests/foreman/cli/test_activationkey.py", line 511, in test_negative_update_description
    u'organization-id': self.org['id'],
AssertionError: CLIReturnCodeError not raised
----------------------------------------------------------------------
Ran 47 tests in 2138.697s

FAILED (SKIP=3, errors=1, failures=2)
```

1 Error is raised due 1180282 BZ which is VERIFIED in Upstream
2 Failures are raised due 1177158 BZ which is also VERIFIED in Upstream

sthirugn commented:
`1 Error is raised due 1180282 BZ which is VERIFIED in Upstream`
`2 Failures are raised due 1177158 BZ which is also VERIFIED in Upstream`
If you dont know already, these can be avoided by `upstream=false` in properties.

oshtaier commented:
@sthirugn Wow. Thank you so much for that information. I completely missed that fact

abalakh commented:
ACK

oshtaier commented:
```
Ran 47 tests in 1985.094s

OK (SKIP=6)
```

oshtaier commented:
@sthirugn @elyezer @sghai @jyejare If you have any spare moment to review that PR I will appreciate  that as tomorrow my (actually, not onlt mine) last day before long PTO.

As usually, next URL can be used for review as github doesn't show big diffs for one file:
https://github.com/oshtaier/robottelo/commit/f97f47b1306eaaa72ff82834a287bd773a222c48.diff

oshtaier commented:
And let's know whether any of your comment is blocking and whether fix for it should be confirmed by you personally

sthirugn commented:
I am starting to look at this now.

oshtaier commented:
@sthirugn Thank you!!!

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -1784,6 +1783,8 @@ def setup_org_for_a_custom_repo(options=None):
                                     if empty)
+    :return: A dictionary representing all entities ids used in our setup
+        procedure (like content view or activation key id)
 
```
To make this more clear, may be
`A dictionary with the entity ids of Activation key, Content view, LifeCycle Environment, Organization, Product and Repository`

sthirugn replied:
also this blank line can be removed.

oshtaier replied:
Sure. I will change text of the message

oshtaier replied:
No. Actually, you cannot remove that line:
```
robottelo/robottelo/cli/factory.py:docstring of robottelo.cli.factory.setup_org_for_a_custom_repo:25: WARNING: Literal block ends without a blank line; unexpected unindent.
```

sthirugn commented on robottelo/cli/factory.py:
```diff
@@ -1910,6 +1918,9 @@ def setup_org_for_a_rh_repo(options=None):
 
+    :return: A dictionary representing all entities ids used in our setup
+        procedure (like content view or activation key id)
+
```
Same two comments as above

sthirugn commented:
Any reason why this was removed? 
```-    def test_positive_create_with_host_collection(self):```

sthirugn commented:
```     def test_positive_create_with_usage_limit_default(self):
...
+        new_ak = self._make_activation_key({u'unlimited-content-hosts': '1'})
+        self.assertEqual(new_ak['content-host-limit'], u'Unlimited')
```
Since it says default, I should get the same result without passing `u'unlimited-content-hosts': '1'`.  Are we checking this too? If not can we add a test or update this test to cover that scenario as well?

oshtaier commented:
`def test_positive_create_with_host_collection(self):` was removed, because you can associate activation key with host collection using `add-host-collection` command only, so it doesn't make sense to have `....create....` method anymore and have only `....update....` left.

oshtaier commented:
Sure I will add one more to test that thing explicitly:
```
    @tier1
    def test_positive_create_with_usage_limit_default(self):
        """@Test: Create Activation key with default Usage limit (Unlimited)

        @Feature: Activation key - Positive Create

        @Assert: Activation key is created
        """
        new_ak = self._make_activation_key()
        self.assertEqual(new_ak['content-host-limit'], u'Unlimited')
```
```
    @tier1
    def test_positive_create_with_usage_limit_default_explicitly(self):
        """@Test: Create Activation key with default Usage limit (Unlimited)

        @Feature: Activation key - Positive Create

        @Assert: Activation key is created
        """
        new_ak = self._make_activation_key({u'unlimited-content-hosts': '1'})
        self.assertEqual(new_ak['content-host-limit'], u'Unlimited')
```

sthirugn commented:
I see that the following test is replaced
```
-    @stubbed()
-    def test_positive_delete_manifest(self):
-        """@Test: Check if deleting a manifest removes it from Activation key
```
to
```
+    def test_positive_delete_subscription(self):
+        """@Test: Check if deleting a subscription removes it from Activation key
```

I like to keep it with old name and test description.  Here we are actually removing the manifest without removing the subscription from activation key.  The old name and description clearly explains that.  Can you please revert it?

oshtaier commented:
We cannot delete manifest from activation key as far as I know, so I implemented the same approach used in webui. We delete subscription from ak, not the manifest.

sthirugn commented:
Yes the implementation is incorrect and needs update - may be create a separate issue to deal with this later. But this is what was intended:
1. Upload manifest
2. Create activation key - attach some subscriptions
3. Delete manifest
4. See if the activation key automatically removed the subscriptions.

oshtaier commented:
@sthirugn got it. I look more accurately and you are right. My bad

sthirugn commented:
No worries, perhaps leave this as stubbed so you can fix it later

oshtaier commented:
@sthirugn but I like to leave that test case as it is as it makes sense. Let me back stub back and leave that one here

sthirugn commented:
sounds great

oshtaier commented:
Returned stubbed test case and used @sthirugn steps as they sound better:
```
    def test_positive_delete_manifest(self):
        """@Test: Check if deleting a manifest removes it from Activation key
        
        @Feature: Activation key - Manifest
        
        @Steps:
        1. Upload manifest
        2. Create activation key - attach some subscriptions
        3. Delete manifest
        4. See if the activation key automatically removed the subscriptions.
        
        @Assert: Deleting a manifest removes it from the Activation key
        
        @Status: Manual
        """
```

oshtaier commented:
```
nosetests tests/foreman/cli/test_activationkey.py -m test_positive_create_with_usage_limit_default
..
----------------------------------------------------------------------
Ran 2 tests in 17.975s

OK
```

sthirugn commented:
```
-    @stubbed()
-    def test_positive_update_aks_to_chost(self):
+    @skip_if_not_set('clients')
+    @tier3
+    def test_positive_add_multiple_to_chost(self):
```
Two things:
1. I like to keep the old test name (more explanation below)
2. The implementation is incorrect.  You can keep this and add a new one or update this one with the explanation below.

This is what was intended:
You can register a content host with more than one activation key.  So if you go to the content host, you should register with multiple activation keys in one command.
```
# subscription-manager register --help
....
  --activationkey=ACTIVATION_KEYS
                        activation key to use for registration (can be
                        specified more than once)
...
...
```

sthirugn commented:
```
   @tier1
    def test_negative_create_with_usage_limit(self):
        """@Test: Create Activation key with invalid Usage Limit

        @Feature: Activation key - Negative Create

        @Assert: Activation key is not created. Appropriate error shown.
        """
        for limit in invalid_values_list():
            with self.subTest(limit):
                with self.assertRaises(CLIFactoryError):
                    self._make_activation_key({
                        u'unlimited-content-hosts': '0',
                        u'max-content-hosts': limit,
                    })
```
Invalid limit should also check for some invalid numbers

sthirugn commented:
```
 @run_only_on('sat')
    @tier2
    def test_positive_update_lce(self):
        """@Test: Update Environment in an Activation key

        @Feature: Activation key - Positive Update

        @Assert: Activation key is updated
        """
        ak_env = self._make_activation_key({
            u'lifecycle-environment-id': self.get_default_env()['id'],
        })
        env = make_lifecycle_environment({u'organization-id': self.org['id']})
        new_cv = make_content_view({u'organization-id': self.org['id']})
        ContentView.publish({u'id': new_cv['id']})
        cvv = ContentView.info({u'id': new_cv['id']})['versions'][0]
        ContentView.version_promote({
            u'id': cvv['id'],
            u'to-lifecycle-environment-id': env['id'],
        })
        ActivationKey.update({
            u'id': ak_env['id'],
            u'lifecycle-environment-id': env['id'],
            u'content-view': new_cv['name'],
            u'organization-id': self.org['id'],
        })
        updated_ak = ActivationKey.info({'id': ak_env['id']})
        self.assertEqual(updated_ak['lifecycle-environment'], env['name'])
```
Is there any reason you are publish/promoting just for a lce update?  If no, we can remove those steps and just update lce?

sthirugn commented:
```
@tier1
    def test_positive_update_usage_limit_to_finite_number(self):
        """@Test: Update Usage limit from Unlimited to a finite number

        @Feature: Activation key - Positive Update

        @Assert: Activation key is updated
        """
        new_ak = self._make_activation_key()
        self.assertEqual(new_ak['content-host-limit'], u'Unlimited')
        ActivationKey.update({
            u'unlimited-content-hosts': '0',
            u'max-content-hosts': '10',
```
May be we should check the max allowed number here and use a random number of certain range to get more coverage?

sthirugn commented:
Great work. ACK pending comments

oshtaier commented:
Changed name to old one. Added new stub:
```
    @skip_if_not_set('clients')
    @stubbed()
    @tier3
    def test_positive_update_aks_to_chost_in_one_command(self):
        """@Test: Check if multiple Activation keys can be attached to a
        Content host in one command. Here is a command details:
        
        --activationkey=ACTIVATION_KEYS activation key to use for registration
        (can be specified more than once)
        
        That means that we can re-use same command few times to add different
        keys

        @Feature: Activation key - Content host

        @Assert: Multiple Activation keys are attached to a Content host
        """
```

oshtaier commented:
Added invalid numbers for limit variable:
```
def test_negative_create_with_usage_limit(self):
        """@Test: Create Activation key with invalid Usage Limit

        @Feature: Activation key - Negative Create

        @Assert: Activation key is not created. Appropriate error shown.
        """
        include_list = ['-1', '-500', 0, 0.5]
        for limit in invalid_values_list() + include_list:
            with self.subTest(limit):
                with self.assertRaises(CLIFactoryError):
                    self._make_activation_key({
                        u'unlimited-content-hosts': '0',
                        u'max-content-hosts': limit,
                    })
```
```
nosetests tests/foreman/cli/test_activationkey.py -m test_negative_create_with_usage_limit
.
----------------------------------------------------------------------
Ran 1 test in 42.918s

OK
```

oshtaier commented:
```
env = make_lifecycle_environment({u'organization-id': self.org['id']})
new_ak_env = self._make_activation_key({
    u'lifecycle-environment-id': env['id'],
})
```
Works properly
```
ak_env = self._make_activation_key({
    u'lifecycle-environment-id': self.get_default_env()['id'],
})
env = make_lifecycle_environment({u'organization-id': self.org['id']})
ActivationKey.update({
    u'id': ak_env['id'],
    u'lifecycle-environment-id': env['id'],
    u'organization-id': self.org['id'],
})
```
Will not as it will search for lce in same content view and that definitely give us exception:
```
CLIReturnCodeError: Command "activation-key update" finished with return_code 65
stderr contains following message:
[ERROR 2015-12-23 12:42:46 Exception] Validation failed: Content view 'Default Organization View' is not in environment 'Fr20mE'
Could not update the activation key:
  Validation failed: Content view 'Default Organization View' is not in environment 'Fr20mE'
```
So it is necessary to have proper association to be able perform update operation

oshtaier commented:
Changed to:
```
    @tier1
    def test_positive_update_usage_limit_to_finite_number(self):
        """@Test: Update Usage limit from Unlimited to a finite number

        @Feature: Activation key - Positive Update

        @Assert: Activation key is updated
        """
        new_ak = self._make_activation_key()
        self.assertEqual(new_ak['content-host-limit'], u'Unlimited')
        ActivationKey.update({
            u'unlimited-content-hosts': '0',
            u'max-content-hosts': '2147483647',
            u'name': new_ak['name'],
            u'organization-id': self.org['id'],
        })
        updated_ak = ActivationKey.info({'id': new_ak['id']})
        self.assertEqual(updated_ak['content-host-limit'], u'2147483647')
```
In my opinion it is really nice to verify boundary value all the time.

sthirugn commented:
yep +1 if you like to add specific boundary tests

oshtaier commented:
```
nosetests tests/foreman/cli/test_activationkey.py -m test_positive_update_usage_limit_to_finite_number
.
----------------------------------------------------------------------
Ran 1 test in 19.475s

OK
```

oshtaier commented:
Addressed all comments. Of course, I will like to avoid these two stubbs, but @sthirugn comments revealed some facts that will require some rework of framework and some more investigation that will require few more efforts that are out of scope of current PR.

Thanks @sthirugn for your knowledge in this are and provided comments

oshtaier commented:
@sthirugn Please, let me know if I need to change something otherwise let's merge this PR. Anyway, I will stay late today to finish it (bad habit to leave something in old year :))

sthirugn commented:
@oshtaier yw, Glad that I am able to help.  Looking again now.

sthirugn commented:
@oshtaier ty for quickly addressing the comments.  I have the just one minor comment:

```
 That means that we can re-use same command few times to add different keys
```
please change to:
```
 This means that we can re-use `--activationkey` option more than once to add different keys
```

oshtaier commented:
Thanks. Changed it according your suggestion

sthirugn commented:
APT
